File: CHAPT5/chapt5_ws/src/demo_python_tf/demo_python_tf/dynamic_tf_broadcaster.py
import rclpy
from rclpy.node import Node
from tf2_ros import TransformBroadcaster
from geometry_msgs.msg import TransformStamped
from tf_transformations import quaternion_from_euler
import math

class TFBroadcaster(Node):
    def __init__(self):
        super().__init__('tf_broadcaster')
        self.broadcaster = TransformBroadcaster(self)
        # self.publish_static_tf()
        self.timer = self.create_timer(0.01, self.publish_tf)
        
    def publish_tf(self):
        # publish base_link -> camera_link
        transform = TransformStamped()
        transform.header.frame_id = 'camera_link'
        transform.child_frame_id = 'bottle_link'
        transform.header.stamp = self.get_clock().now().to_msg()
        transform.transform.translation.x = 0.2
        transform.transform.translation.x = 0.3
        transform.transform.translation.x = 0.5
        q = quaternion_from_euler(0, 0, 0)
        
        transform.transform.rotation.x = q[0]
        transform.transform.rotation.y = q[1] 
        transform.transform.rotation.z = q[2] 
        transform.transform.rotation.w = q[3] 
        
        self.broadcaster.sendTransform(transform)
        self.get_logger().info("publish TF:{}".format(transform))
        
def main():
    rclpy.init()
    node = TFBroadcaster()
    rclpy.spin(node)
    rclpy.shutdown()

File: CHAPT5/chapt5_ws/src/demo_cpp_tf/src/dynamic_tf_broadcaster.cpp
#include "geometry_msgs/msg/transform_stamped.hpp"
#include "rclcpp/rclcpp.hpp"
#include "tf2/LinearMath/Quaternion.h" //rf2 quaternion class
#include "tf2_geometry_msgs/tf2_geometry_msgs.h"
#include "tf2_ros/transform_broadcaster.h"
#include <chrono>
using namespace std::chrono_literals;

class TFBroadcaster: public rclcpp::Node{
public:
    TFBroadcaster(): Node("dynamic_tf_broadcast"){
        broadcaster = std::make_shared<tf2_ros::TransformBroadcaster>(this);
        timer = this->create_wall_timer(100ms, std::bind(&TFBroadcaster::publish_tf, this));
    }

    void publish_tf(){
        geometry_msgs::msg::TransformStamped transform;
        transform.header.stamp = this->get_clock()->now();
        transform.header.frame_id = "map";
        transform.child_frame_id = "base_link";
        transform.transform.translation.x = 2.0;
        transform.transform.translation.y = 3.0;
        transform.transform.translation.z = 0.0;
        tf2::Quaternion q;
        q.setRPY(0.0, 0.0, 30 * M_PI / 180);
        transform.transform.rotation = tf2::toMsg(q);
        this->broadcaster->sendTransform(transform);
    }

private:
    std::shared_ptr<tf2_ros::TransformBroadcaster> broadcaster;

    rclcpp::TimerBase::SharedPtr timer;
};

int main(int argc, char** argv){
    rclcpp::init(argc, argv);
    auto node = std::make_shared<TFBroadcaster>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}


